### PR TITLE
Compile fail stable

### PR DESCRIPTION
**Pull request by GuillaumeGomez**
Since #30726, we never made the `compile_fail` flag nor the error code check stable. I think it's time to change this fact.

r? @alexcrichton

### CODE REVIEW

**Comment by Gankro** [comment]
I'm a big fan of having "this doesn't compile" on stable, but as per our historical gripes with compile-fail tests, I'm not sure if we should allow stable code to check error codes, since I'm like 99.9% sure those aren't stable (in that something may stop compiling for a different reason).

**Comment by GuillaumeGomez** [comment]
If there is a consensus for not stabilizing the error code check, I'll just remove the second commit.

**Comment by alexcrichton** [comment]
Has this been discussed for stabilization? Was this discussed with @rust-lang/dev-tools?

I would personally still be against stabilizing error codes, and do we have sufficient documentation to stabilize this feature as well?

**Comment by kennytm** [comment]
cc #42288 for documentation.

**Comment by GuillaumeGomez** [comment]
Ok so I removed the error codes from the stabilization and added the doc for the `compile_fail` directive (and improved the doc a bit too).

**Comment by alexcrichton** [comment]
Are there any tests for this in the `src/test/rustdoc` folder? I'd expect that if we had existing tests this'd remove the feature gates from those tests.

**Comment by alexcrichton** [comment]
I'll I'm going to defer to @nrc for review as I think the dev-tools team will want to sign off on this.

r? @nrc

**Comment by GuillaumeGomez** [comment]
The tests are literally all over the docs. Where would you want me to add them?

**Comment by alexcrichton** [comment]
Can you add a unit test for this in `src/test/rustdoc` to ensure that it doesn't require a feature gate?

**Comment by GuillaumeGomez** [comment]
Sure.

File: src/doc/rustdoc/src/documentation-tests.md [hunk 1]
```diff
@@ -43,17 +56,17 @@ Sometimes, you need some setup code, or other things that would distract
...
 an example block that looks like this:
 
-```text
-/// Some documentation.
-# fn foo() {}
-```
+   ```text
+   /// Some documentation.
+   # fn foo() {}
+   ```
```
Comment (frewsxcv) [inline src/doc/rustdoc/src/documentation-tests.md]: why did these get indented?

Comment (GuillaumeGomez) [inline src/doc/rustdoc/src/documentation-tests.md]: To be able to see the backticks when rendered.

**Comment by nrc** [comment]
I don't think we can yet get an auto-rfc-bot thing for the dev-tools/docs team, so manually. I propose we should accept this (stabilise the flag) - seems like a fairly harmless feature that has not changed in the last 1.5 years and is considered useful by some people.

Please sign off - @michaelwoerister @killercup @QuietMisdreavus @steveklabnik @jonathandturner @japaric

File: src/doc/rustdoc/src/documentation-tests.md (hunk 1 above)
Comment (QuietMisdreavus) [inline src/doc/rustdoc/src/documentation-tests.md]: But that's not the point. The point is to show what happens when you hide a line with `#`, not to highlight the text differently.

Comment (QuietMisdreavus) [inline src/doc/rustdoc/src/documentation-tests.md]: also it caused [a travis failure](https://travis-ci.org/rust-lang/rust/jobs/267150878):

    [01:04:47] ---- /checkout/src/doc/rustdoc/src/documentation-tests.md - Documentation_tests (line 68) stdout ----
    [01:04:47] 	error: unknown start of token: `

**Comment by QuietMisdreavus** [comment]
I approve this change.

In the docs team meeting yesterday, we mentioned that it might be prudent to include a note in the documentation stating that "this code doesn't compile" isn't guaranteed to be stable from release to release. Language features like NLL, struct field init shorthand, using a nightly feature without a feature flag (if it's stabilized as-is), etc, can make code that was previously broken start compiling. Making it explicit that some things may change between releases would be helpful in terms of cutting down on "where is this documented" questions.

**Comment by GuillaumeGomez** [comment]
I *think* I made the changes you required @QuietMisdreavus. Or did I miss anything?

File: src/doc/rustdoc/src/documentation-tests.md [hunk 2]
```diff
@@ -136,7 +149,7 @@ explanation.
 Another case where the use of `#` is handy is when you want to ignore
 error handling. Lets say you want the following,
 
-```rust,ignore
+```rust
```
Comment (QuietMisdreavus) [inline src/doc/rustdoc/src/documentation-tests.md]: Note that taking the `ignore` off of this keeps making the test fail, because it's a doc comment that doesn't document anything.

File: src/doc/rustdoc/src/documentation-tests.md [hunk 3]
```diff
@@ -233,6 +246,16 @@ not actually pass as a test.
 # fn foo() {}
 ```
 
+`compile_fail` tells `rustdoc` that the compilation should fail. If it
+compiles, then the test will fail.
```
Comment (QuietMisdreavus) [inline src/doc/rustdoc/src/documentation-tests.md]: Still looking for a note here about how things may start compiling in later releases.

File: src/doc/rustdoc/src/documentation-tests.md [hunk 4]
```diff
@@ -16,6 +16,19 @@ The basic idea is this:
 The triple backticks start and end code blocks. If this were in a file named `foo.rs`,
 running `rustdoc --test foo.rs` will extract this example, and then run it as a test.
 
+Please note that by default, if no language is set for the block code, `rustdoc`
+assumes it is `Rust` code. So the following:
```
Comment (QuietMisdreavus) [inline src/doc/rustdoc/src/documentation-tests.md]: This block doesn't seem related to the rest of this PR? Like, it's worth noting, but it's not part of "making `compile_fail` stable".

Comment (GuillaumeGomez) [reply src/doc/rustdoc/src/documentation-tests.md]: Do you want me to put this change in another commit?

File: src/doc/rustdoc/src/documentation-tests.md (hunk 2 above)
Comment (GuillaumeGomez) [reply src/doc/rustdoc/src/documentation-tests.md]: Erf, good point!

File: src/doc/rustdoc/src/documentation-tests.md [hunk 5]
```diff
@@ -233,6 +246,17 @@ not actually pass as a test.
 # fn foo() {}
 ```
 
+`compile_fail` tells `rustdoc` that the compilation should fail. If it
+compiles, then the test will fail. However please note that a code
+failing with the current rustc version may works in a future release.
```
Comment (QuietMisdreavus) [inline src/doc/rustdoc/src/documentation-tests.md]: Light phrasing nit: "However please note that code failing with the current Rust release may work in a future release, as new features are added."

File: src/doc/rustdoc/src/documentation-tests.md [hunk 6]
```diff
@@ -16,6 +16,19 @@ The basic idea is this:
 The triple backticks start and end code blocks. If this were in a file named `foo.rs`,
 running `rustdoc --test foo.rs` will extract this example, and then run it as a test.
 
+Please note that by default, if no language is set for the block code, `rustdoc`
+assumes it is `Rust` code. So the following:
+
+    ```rust
+    let x = 5;
+    ```
```
Comment (QuietMisdreavus) [inline src/doc/rustdoc/src/documentation-tests.md]: Don't indent these docblocks. It's treating the ``` as part of the doctest and failing to parse it.

File: src/doc/rustdoc/src/documentation-tests.md [hunk 7]
```diff
@@ -233,6 +246,18 @@ not actually pass as a test.
...
+compiles, then the test will fail. However please note that code failing
+with the current Rust release may work in a future release, as new features
+are added.
+
+```rust
+/// ```compile_fail
+/// let x = 5;
+/// x += 2; // shouldn't compile!
+/// ```
+```
```
Comment (QuietMisdreavus) [inline src/doc/rustdoc/src/documentation-tests.md]: "Documentation comment that doesn't document anything"

Probably good to stick `ignore` on this one.

**Review by QuietMisdreavus — COMMENTED** [review]
More travis failures

**Comment by GuillaumeGomez** [comment]
Finally all good!

**Comment by GuillaumeGomez** [comment]
Any news in here?

**Comment by carols10cents** [comment]
ping the @rust-lang/dev-tools team for signoffs; here's some manual ticky boxes for you:

- [ ] @nrc 
- [ ] @michaelwoerister 
- [ ] @killercup 
- [ ] @QuietMisdreavus 
- [ ] @steveklabnik 
- [ ] @jonathandturner 
- [ ] @japaric

**Comment by killercup** [comment]
Assuming all the stuff from https://github.com/rust-lang/rust/pull/43949#issuecomment-324648649 is addressed (it appears so), LGTM.

@carolbot reviewed

**Comment by michaelwoerister** [comment]
@rfcbot reviewed
